apixu-weatherbot1.0.tcl

```diff
@@ -231,69 +233,86 @@ namespace eval weather {
         #   type(string) - Current weather or forecast called.
         #
         # Returns:
-        #   (string) Returns json data received from the api.
+        #   (dict) Returns json data received from the api and
+        #          converted to a dict.
         #
-        putlog "weather::_get_json was called"
-
-        # First regexp matches any location given by user that is not
-        # a-z, A-Z, number, period, comma or space. This is sanitize
-        # the location string given to catch any illegal characters
-        # before the API call that will return an error.
-        #
-        # Second regexp matches any location give by user that starts
-        # and ends with a-z or A-Z with a : in the between the two. This
-        # is to let strings like "iata:ksjc" to be passed through as the
-        # only illegal character if used the correct way.
-        #
-        # If the first regexp is true and second regexp is false, the
-        # conditional will be true and throw an error for illegal characters.
-        if {[regexp {[^\w., ]} $location] && ![regexp {^[a-zA-Z]+:[a-zA-Z]+$} $location]} {
-            putlog "weather::_get_json caught illegal characters"
-            error "Illegal characters in your location query. Acceptable characters:\
-                   a-z, 0-9, commas, and spaces. Colons only used i.e. 'iata:ksjc'"
+        putlog "weather::_get_weather_datawas called"
+        set query [::http::formatQuery key $::weather::apikey q $location]
+        set url "$::weather::base_url/$type.json?$query"
+        if {$type ne "current"} {
+            set url "$::weather::base_url/$type.json?$query&days=$::weather::forecast_days"
         }
 
-        # Replace all spaces in $location with %20 for theapi link call
-        regsub -all -- { } $location {%20} location
+        putlog "weather::_get_weather_data getting data from $url"
+        set token [::http::geturl $url -timeout 750]
+        set status [::http::status $token]
+
+        # If a timeout occurs to the api request, retry up to
+        # max_retries with backoff. Default backoff set to 1000ms.
+        if {$status eq "timeout"} {
+            ::http::cleanup $token
+            set retry 1
+            set max_retries 10
+            set backoff 1000
 
-        set url "$::weather::base_url/$type.json?key=$::weather::apikey&q=$location"
-        if {$type ne "current"} {
-            set url "$::weather::base_url/$type.json?key=$::weather::apikey&q=$location&days=5"
-        }
+            while {$retry <= $max_retries} {
+                set token [::http::geturl $url -timeout 750]
+                set status [::http::status $token]
 
-        putlog "weather::_get_json getting data from $url"
-        if {[catch {set token [::http::geturl $url -timeout 10000]} errormsg]} {
-            putlog "::http::geturl Error: $errormsg"
-            error "No matching location found"
+                if {$retry eq 3} {
```

Comment:
instead of `eq`, maybe do `if {($retry % 3) eq 0}` in order to notify the user every 3 retries.  also make the notification interval configurable.